appinterface/projection/crossfire/crossfire.go

```diff
@@ -0,0 +1,754 @@
+	}
+
+	committed := false
+	defer func() {
+		if !committed {
+			_ = rdbTx.Rollback()
+		}
+	}()
+
+	rdbTxHandle := rdbTx.ToHandle()
+	crossfireValidatorsView := view.NewCrossfireValidators(rdbTxHandle)
+	crossfireChainStatsView := view.NewCrossfireChainStats(rdbTxHandle)
+	crossfireValidatorsStatsView := view.NewCrossfireValidatorsStats(rdbTxHandle)
+
+	// Event projection
+	var blockTime utctime.UTCTime
+	for _, event := range events {
+		if blockCreatedEvent, ok := event.(*event_usecase.BlockCreated); ok {
+			blockTime = blockCreatedEvent.Block.Time
+			if handleErr := projection.handleBlockCreatedEvent(crossfireChainStatsView, crossfireValidatorsView, crossfireValidatorsStatsView, blockCreatedEvent); handleErr != nil {
+				return fmt.Errorf("error handling BlockCreatedEvent: %v", handleErr)
+			}
+		}
+	}
+
+	if err := projection.projectCrossfireValidatorView(crossfireValidatorsView, crossfireChainStatsView, crossfireValidatorsStatsView, height, blockTime, events); err != nil {
+		return fmt.Errorf("error projecting validator view: %v", err)
+	}
+
+	// Ranks computation at the end of event projection
+	if blockTime.Before(projection.competitionEndTime) {
+		if err := projection.computeTxSentRank(crossfireValidatorsStatsView, crossfireValidatorsView); err != nil {
+			return fmt.Errorf("error Updating TxSentTask Rank %v", err)
+		}
+	}
+	if blockTime.Before(projection.phaseThreeStartTime) {
+		if err := projection.computeCommitmentRank(
+			"rank_task_phase_1_2_commitment_count",
+			constants.PHASE_1N2_COMMIT_PREFIX,
+			crossfireValidatorsStatsView,
+			crossfireValidatorsView,
+		); err != nil {
+			return fmt.Errorf("error compute rank_task_phase_1_2_commitment_count %v", err)
+		}
+	}
+	if blockTime.After(projection.phaseThreeStartTime) && blockTime.Before(projection.competitionEndTime) {
+		if err := projection.computeCommitmentRank(
+			"rank_task_phase_3_commitment_count",
+			constants.PHASE_3_COMMIT_PREFIX,
+			crossfireValidatorsStatsView,
+			crossfireValidatorsView,
+		); err != nil {
+			return fmt.Errorf("error compute rank_task_phase_3_commitment_count %v", err)
+		}
+	}
+
+	// Done current height projection and ranking computation
+	if err := projection.UpdateLastHandledEventHeight(rdbTxHandle, height); err != nil {
+		return fmt.Errorf("error updating last handled event height: %v", err)
+	}
+
+	if err := rdbTx.Commit(); err != nil {
+		return fmt.Errorf("error committing changes: %v", err)
+	}
+	committed = true
+	return nil
+}
+
+func (projection *Crossfire) handleBlockCreatedEvent(
+	crossfireChainStatsView *view.CrossfireChainStats,
+	crossfireValidatorsView *view.CrossfireValidators,
+	crossfireValidatorsStatsView *view.CrossfireValidatorsStats,
+	event *event_usecase.BlockCreated,
+) error {
+	blockTime := event.Block.Time
+	blockHeight := event.Block.Height
+
+	// increment current phase block number
+	if blockTime.After(projection.phaseOneStartTime) && blockTime.Before(projection.phaseTwoStartTime) {
```

Comment:
Missing `Equal` check here?

Reply:
The After check is actually `>=` so actually it is `AtOrAfter`